File: common/clock_tree.py
#!/usr/bin/env python3
# -*- coding:utf-8 -*-

# Licensed under the GNU General Public License v. 3 (the "License")
# You may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.gnu.org/licenses/gpl-3.0.html
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# Copyright (C) 2022-2024 xqyjlj<xqyjlj@126.com>
#
# @author      xqyjlj
# @file        clock_tree.py
#
# Change Logs:
# Date           Author       Notes
# ------------   ----------   -----------------------------------------------
# 2024-10-28     xqyjlj       initial version
#

import json
import os

import jsonschema
import yaml
from loguru import logger

from .i18n_type import I18nType
from .settings import SETTINGS


class ClockTreeType:
    class ElementUnitType:
        def __init__(self, data: dict):
            self.__data = data

        def __str__(self) -> str:
            return json.dumps(self.__data, indent=2, ensure_ascii=False)

        @property
        def origin(self) -> dict:
            return self.__data

        @property
        def refParameter(self) -> str:
            return self.__data.get("refParameter", '')

        @property
        def type(self) -> str:
            return self.__data.get("type", '')

        @property
        def enable(self) -> str | bool | None:
            return self.__data.get("enable", None)

        @property
        def output(self) -> list[str]:
            return self.__data.get("output", [])

        @property
        def input(self) -> list[str]:
            return self.__data.get("input", [])

        @property
        def z(self) -> int:
            return self.__data.get("z", -1)

    def __init__(self, data: dict):
        self.__data = data
        self.__elements = None
        self.__i18n = None

    def __str__(self) -> str:
        return json.dumps(self.__data, indent=2, ensure_ascii=False)

    @property
    def origin(self) -> dict:
        return self.__data

    @property
    def elements(self) -> dict[str, ElementUnitType]:
        if self.__elements is None:
            self.__elements = {}
            for name, element in self.__data.get('elements', {}).items():
                self.__elements[name] = self.ElementUnitType(element)
        return self.__elements

    @property
    def i18n(self) -> dict[str, I18nType]:
        if self.__i18n is None:
            self.__i18n = {}
            for name, i in self.__data.get('i18n', {}).items():
                self.__i18n[name] = I18nType(i)
        return self.__i18n

    def i18nOrigin(self) -> dict[str, I18nType]:
        return self.__data.get('i18n', {})


class ClockTree:

    def __init__(self):
        self.__clockTrees = {}

    @logger.catch(default=False)
    def __checkClockTree(self, clockTree: dict) -> bool:
        with open(os.path.join(SETTINGS.DATABASE_FOLDER, "schema", "clock_tree.yml"), 'r', encoding='utf-8') as f:
            schema = yaml.load(f.read(), Loader=yaml.FullLoader)
            jsonschema.validate(instance=clockTree, schema=schema)
        return True

    @logger.catch(default=ClockTreeType({}))
    def __getClockTree(self, vendor: str, name: str) -> ClockTreeType:
        file = os.path.join(SETTINGS.DATABASE_FOLDER, "clock", vendor.lower(), f"{name.lower()}.yml")
        if os.path.isfile(file):
            with open(file, 'r', encoding='utf-8') as f:
                clockTree = yaml.load(f.read(), Loader=yaml.FullLoader)
                succeed = self.__checkClockTree(clockTree)
            if succeed:
                return ClockTreeType(clockTree)
            else:
                return ClockTreeType({})
        else:
            logger.error(f"{file} is not file!")
            return ClockTreeType({})

    def getClockTree(self, vendor: str, name: str) -> ClockTreeType:
        vendor = vendor.lower()
        name = name.lower()
        if vendor in self.__clockTrees and name in self.__clockTrees[vendor]:
            return self.__clockTrees[vendor][name]
        else:
            # noinspection PyTypeChecker,PyArgumentList
            clockTree = self.__getClockTree(vendor, name)
            if vendor not in self.__clockTrees:
                self.__clockTrees[vendor] = {}
            self.__clockTrees[vendor][name] = clockTree
            # noinspection PyTypeChecker
            return clockTree

    def clockTrees(self) -> dict[str, ClockTreeType]:
        return self.__clockTrees


CLOCK_TREE = ClockTree()

File: src/database/clockTree.ts
/**
 * ****************************************************************************
 *  @author      xqyjlj
 *  @file        clockTree.ts
 *  @brief
 *
 * ****************************************************************************
 *  @attention
 *  Licensed under the Apache License v. 2 (the "License");
 *  You may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0.html
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *  Copyright (C) 2025-2025 xqyjlj<xqyjlj@126.com>
 *
 * ****************************************************************************
 *  Change Logs:
 *  Date           Author       Notes
 *  ------------   ----------   -----------------------------------------------
 *  2025-05-15     xqyjlj       initial version
 */

import type {
  ClockTreeEdgeType,
  ClockTreeEdgeTypeType,
  ClockTreeNodeType,
  ClockTreeNodeTypeType,
  ClockTreeNodeUnitType,
  ClockTreeType,
} from '@/electron/types'
import type { XYPosition } from '@vue-flow/core'
import type { App } from 'vue'
import type { Ip, IpManager, IpParameter } from './ip'
import { computed, inject, markRaw } from 'vue'

// #region typedef

export class ClockTree {
  private _nodes: Record<string, ClockTreeNode> = {}
  private _edges: Record<string, ClockTreeEdge> = {}

  constructor(
    private _origin: ClockTreeType,
    public ip: Ip,
  ) {
    for (const [name, node] of Object.entries(this._origin.nodes)) {
      this._nodes[name] = new ClockTreeNode(node, name, this)
    }

    for (const [name, edge] of Object.entries(this._origin.edges)) {
      this._edges[name] = new ClockTreeEdge(edge)
    }
  }

  get origin(): ClockTreeType {
    return this._origin
  }

  get instance(): string {
    return this._origin.instance
  }

  get nodes(): Record<string, ClockTreeNode> {
    return this._nodes
  }

  get edges(): Record<string, ClockTreeEdge> {
    return this._edges
  }

  updateEdges(edges: Record<string, ClockTreeEdgeType>) {
    this._edges = {}
    this._origin.edges = edges
    for (const [name, edge] of Object.entries(this._origin.edges)) {
      this._edges[name] = new ClockTreeEdge(edge)
    }
  }
}

export class ClockTreeEdge {
  constructor(
    private _origin: ClockTreeEdgeType,
  ) {
  }

  get origin(): ClockTreeEdgeType {
    return this._origin
  }

  get type(): ClockTreeEdgeTypeType {
    return this._origin.type
  }

  get source(): string {
    return this._origin.source
  }

  get target(): string {
    return this._origin.target
  }

  get sourceHandle(): string | undefined | null {
    return this._origin.sourceHandle
  }

  get targetHandle(): string | undefined | null {
    return this._origin.targetHandle
  }

  get label(): string | undefined {
    return this._origin.label
  }
}

export class ClockTreeNode {
  constructor(
    private _origin: ClockTreeNodeType,
    private _name: string,
    private _parent: ClockTree,
  ) {
  }

  get origin(): ClockTreeNodeType {
    return this._origin
  }

  get ip(): Ip {
    return this._parent.ip
  }

  get ref(): string {
    return this._name
  }

  parameter = computed((): IpParameter => {
    const paramName = this._name
    const parameter = this.ip.containers.clockTree.refParameters.value[paramName]?.parameter
    if (!parameter) {
      console.error(`Parameter '${paramName}' not found in IP '${this.ip.instance}'.`)
    }
    return parameter.value
  })

  enabled = computed((): boolean => this.parameter.value.enabled.value)

  get type(): ClockTreeNodeTypeType {
    return this._origin.type
  }

  get position(): XYPosition {
    return this._origin.position ?? { x: 0, y: 0 }
  }

  get unit(): ClockTreeNodeUnitType | undefined {
    return this._origin.unit
  }
}

// #endregion

export class ClockTreeManager {
  private _map: Record<string, Record<string, { clockTree: ClockTree, define: string }>> = markRaw({})
  private _ipManager?: IpManager

  constructor() {
  }

  setIpManager(ipManager: IpManager) {
    this._ipManager = markRaw(ipManager)
  }

  async load(vendor: string, name: string, define: string) {
    const content = await window.electron.invoke('database:getClockTree', vendor, define) as ClockTreeType
    if (content) {
      const clockTree = new ClockTree(content, this._ipManager!.getPeripheral(vendor, content.instance)!);
      (this._map[vendor] ??= {})[name] = markRaw({ clockTree, define })
    }
  }

  save(vendor: string, name: string, clockTree?: ClockTree) {
    if (this._map[vendor]?.[name]) {
      const define = this._map[vendor][name].define
      clockTree ??= this._map[vendor][name].clockTree
      window.electron.send('database:setClockTree', vendor, define, clockTree.origin)
    }
  }

  get(vendor: string, name: string): ClockTree | null {
    if (this._map[vendor]?.[name]) {
      return this._map[vendor][name].clockTree
    }

    return null
  }
}

export function createClockTreeManagerPlugin() {
  const manager = new ClockTreeManager()

  return {
    manager,
    plugin: {
      install(app: App) {
        app.provide('database@clockTreeManager', manager)
      },
    },
    setIpManager(ipManager: IpManager) {
      manager.setIpManager(ipManager)
    },
  }
}

export function useClockTreeManager(): ClockTreeManager {
  return inject('database@clockTreeManager')!
}
